### PR TITLE
Remove caching of checkpoints from the Inbox

### PR DESCRIPTION
**Pull request by asaj**
This PR removes cached checkpoints from `Inbox`. Instead, checkpoints are expected to be verified in the same transaction as message processing.

Rationale:
1. With the move to fraud proofs that require merkle proofs in #475, there is a vulnerability with cached checkpoints where validators cache a fraudulent checkpoint but do not attempt to process messages against it until they're un-bonded. This is easily mitigated by requiring a proof against that checkpoint, but that is counter-productive to the idea of cached checkpoints as a way to save gas.
2. Removing the notion of cached checkpoints reduces complexity of the off-chain agents.
3. Gas savings can still be achieved by batch-processing multiple messages against the same checkpoint.

Other changes:
- Remove `leafIndex` and `proof` from the `Inbox.Process` event, and the unused `InboxValidatorManager.Quorum` event. These were included to allow fraud proofs to be constructed more easily, but it doesn't make sense to incur the extra gas costs until validators are bonded
- Rename `BadRecipientHandle` to `BadRecipient2` for consistency
- Remove redundant `destination` and `messageHash` from `Dispatch` event

### CODE REVIEW

File: solidity/core/contracts/Inbox.sol [hunk 1]
```diff
@@ -81,45 +75,31 @@ contract Inbox is IInbox, Version0, Mailbox {
...
     /**
      * @notice Attempts to process the provided formatted `message`. Performs
      * verification against root of the proof
+     * @dev Called by the validator manager, which is responsible for verifying a
+     * quorum of validator signatures on the checkpoint.
      * @dev Reverts if verification of the message fails.
-     * @dev Includes the eventual function signature for Sovereign Consensus,
-     * but comments out the name to suppress compiler warning
+     * @param _root The merkle root of the checkpoint used to prove message inclusion.
+     * @param _index The index of the checkpoint used to prove message inclusion.
```
Comment (yorhodes) [inline solidity/core/contracts/Inbox.sol:85]: `checkpointIndex`?

[the remark this answers is not in the document]
Comment (asaj) [reply solidity/core/contracts/Inbox.sol:85]: Does that imply you're asking for `checkpointRoot` as well?

File: solidity/core/contracts/Inbox.sol [hunk 2]
```diff
@@ -81,45 +75,31 @@ contract Inbox is IInbox, Version0, Mailbox {
...
     function process(
+        bytes32 _root,
+        uint256 _index,
         bytes calldata _message,
         bytes32[32] calldata _proof,
-        uint256 _index,
-        bytes calldata /* _sovereignData */
-    ) external override {
+        uint256 _leafIndex
+    ) external override onlyValidatorManager {
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/core/contracts/Inbox.sol:96]: As discussed offline, we do not expect most applications to interface with the `InboxValidatorManager`

File: solidity/core/contracts/validator-manager/InboxValidatorManager.sol [hunk 3]
```diff
@@ -12,18 +12,6 @@ import {MultisigValidatorManager} from "./MultisigValidatorManager.sol";
  * them to an Inbox.
  */
 contract InboxValidatorManager is MultisigValidatorManager {
-    // ============ Events ============
-
-    /**
-     * @notice Emitted when a checkpoint has been signed by a quorum
-     * of validators and cached on an Inbox.
-     * @dev This event allows watchers to observe the signatures they need
-     * to prove fraud on the Outbox.
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/core/contracts/validator-manager/InboxValidatorManager.sol:21]: Per our offline discussion, they can not be observed. Given that:
1. The validators are not bonded
2. The watcher code isn't written

it didn't make sense to pay the gas costs of emitting the signatures for now

File: solidity/core/contracts/test/TestValidatorManager.sol [hunk 4]
```diff
@@ -8,11 +8,14 @@ import {IInbox} from "../../interfaces/IInbox.sol";
  * to be a contract.
  */
 contract TestValidatorManager {
-    function cacheCheckpoint(
+    function process(
         IInbox _inbox,
         bytes32 _root,
-        uint256 _index
+        uint256 _index,
+        bytes calldata _message,
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/core/contracts/test/TestValidatorManager.sol:15]: Seemingly it does, 89688 with it last vs 89734 with it as is.

Why did you expect gas savings?

File: solidity/core/test/lib/mailboxes.ts [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+export const dispatchMessage = async (
+  outbox: TestOutbox,
+  destination: number,
+  recipient: string,
+  messageStr: string,
+) => {
+  const tx = await outbox.dispatch(
+    destination,
+    recipient,
+    ethers.utils.formatBytes32String(messageStr),
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/core/test/lib/mailboxes.ts:18]: Agreed

File: solidity/core/test/lib/mailboxes.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  };
+};
+
+export interface MerkleProof {
+  root: string;
+  proof: types.BytesArray;
+  leaf: string;
+  index: BigNumber;
+  message: string;
+}
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/core/test/lib/mailboxes.ts:55]: Import it from where?

File: solidity/core/test/validator-manager/inboxValidatorManager.test.ts [hunk 7]
```diff
@@ -42,42 +46,62 @@ describe('InboxValidatorManager', () => {
...
+        validatorManager.process(
           inbox.address,
-          root,
-          index,
+          proof.root,
+          proof.index,
           signatures,
+          proof.message,
+          proof.proof,
+          proof.index,
```
[the remark this answers is not in the document]
Comment (asaj) [reply solidity/core/test/validator-manager/inboxValidatorManager.test.ts:102]: Yeah I guess the thinking is that the signed checkpoint comes first, and then the proof against that checkpoint

File: typescript/utils/src/types.ts [hunk 8]
```diff
@@ -1,8 +1,44 @@
+import { BytesLike } from 'ethers';
+
 /********* BASIC TYPES *********/
 export type Domain = number;
 export type Address = string;
 export type AddressBytes32 = string;
 export type HexString = string;
+export type BytesArray = [
```
[the remark this answers is not in the document]
Comment (asaj) [reply typescript/utils/src/types.ts:8]: Hmm yeah I was wondering why you removed it. If typechain doesn't play nicely with it, happy to remove

**Pull request merged**